hadoop-hdds/docs/content/design/lifecycle-configurations.md

```diff
@@ -0,0 +1,253 @@
+- Lifecycle configurations (will be stored in DB) consists of volumeName, bucketName and a list of rules
+    - A rule contains prefix (string), Expiration and an optional Filter.
+    - Object tagging integrations for bucket lifecycle configuration.
+    - Expiration contains either days (integer) or Date (long)
+    - Filter contains prefix (string).
+- S3G bucket endpoint needs few updates to accept ?/lifecycle 
+- ClientProtocol and all implementers provides (get, list, delete and create) lifecycle configuration
+- RetentionManager:
+   - Upon startup, the OzoneManager initializes the Retention Manager based on configuration parameters such as retention interval.
+   - A background retention service is responsible for scheduling and executing tasks at specified intervals.
+   - The Retention Manager retrieves lifecycle configurations associated with buckets.
+   - Then assigns each lifecycle configuration (attached to a bucket) to a threadpool (Configurable) for further processing. 
+   - Each task will iterate through keys of a specific bucket and issue deletion request for eligible keys.
+     
+
+
+### Flow
+1. Users interact with lifecycle configurations via S3Gateway.
+2. Configuration details are processed by a handler.
+3. Configurations are saved/fetched from the database.
+4. RetentionManager, running periodically in the Leader OM, executes lifecycle configurations and issues deletions for eligible keys.
+
+## Limitations
+- The current solution lacks certain features:
+  - Only expiration actions are supported.
+  - Lack of CLI support for managing lifecycle configurations across all buckets (S3G is the only supported entry point).
+  
+All these kind of features can be added in the future.
+
+## Protobuf Definitions
+```protobuf
+/**
+S3 lifecycles (filter, expiration, rule and configuration).
+ */
+message LifecycleFilter {
+  optional string prefix = 1;
+}
+
+message LifecycleExpiration {
+  optional uint32 days = 1;
+  optional string date = 2;
+}
+
+message LifecycleRule {
+  optional string id = 1;
+  optional string prefix = 2;
+  required bool enabled = 3;
+  optional LifecycleExpiration expiration = 4;
+  optional LifecycleFilter filter = 5;
+}
+
+message LifecycleConfiguration {
+  required string volume = 1;
+  required string bucket = 2;
+  required string owner = 3;
+  optional uint64 creationTime = 4;
+  repeated LifecycleRule rules = 5;
+  optional uint64 objectID = 6;
+  optional uint64 updateID = 7;
+}
+
+message CreateLifecycleConfigurationRequest {
+  required LifecycleConfiguration lifecycleConfiguration = 1;
+}
+
+message CreateLifecycleConfigurationResponse {
+
+}
+
+message InfoLifecycleConfigurationRequest {
+  required string volumeName = 1;
+  required string bucketName = 2;
+}
+
+message InfoLifecycleConfigurationResponse {
+  required LifecycleConfiguration lifecycleConfiguration = 1;
+}
+
+message DeleteLifecycleConfigurationRequest {
+  required string volumeName = 1;
+  required string bucketName = 2;
```

Comment:
Would accepting an optional LifecycleFilter here and in List and Info configuration requests be useful?

Reply:
I see.